blockchain/blockindex.go

```diff
@@ -440,29 +455,73 @@ func (bi *blockIndex) AddNode(node *blockNode) {
 //
 // This function MUST be called with the block index lock held (for writes).
 func (bi *blockIndex) addChainTip(tip *blockNode) {
-	bi.chainTips[tip.height] = append(bi.chainTips[tip.height], tip)
+	bi.totalTips++
+
+	// When an entry does not already exist for the given tip height, add an
+	// entry to the map with the tip stored in the statically-allocated field.
+	entry, ok := bi.chainTips[tip.height]
+	if !ok {
+		bi.chainTips[tip.height] = chainTipEntry{tip: tip}
+		return
+	}
+
+	// Otherwise, an entry already exists for the given tip height, so store the
+	// tip in the dynamically-allocated slice.
+	entry.otherTips = append(entry.otherTips, tip)
+	bi.chainTips[tip.height] = entry
 }
 
 // removeChainTip removes the passed block node from the available chain tips.
 //
 // This function MUST be called with the block index lock held (for writes).
 func (bi *blockIndex) removeChainTip(tip *blockNode) {
-	nodes := bi.chainTips[tip.height]
-	for i, n := range nodes {
-		if n == tip {
-			copy(nodes[i:], nodes[i+1:])
-			nodes[len(nodes)-1] = nil
-			nodes = nodes[:len(nodes)-1]
-			break
+	// Nothing to do if no tips exist at the given height.
+	entry, ok := bi.chainTips[tip.height]
+	if !ok {
+		return
+	}
+
+	// The most common case is a single tip at the given height, so handle the
+	// case where the tip that is being removed is the tip that is stored in the
+	// statically-allocated field first.
+	if entry.tip == tip {
+		bi.totalTips--
+		entry.tip = nil
+
+		// Remove the map entry altogether if there are no more tips left.
+		if len(entry.otherTips) == 0 {
+			delete(bi.chainTips, tip.height)
+			return
+		}
+
+		// There are still tips stored in the dynamically-allocated slice, so
+		// move the first tip from it to the statically-allocated field, nil the
+		// slice so it can be garbage collected when there are no more items in
+		// it, and update the map with the modified entry accordingly.
+		entry.tip = entry.otherTips[0]
```

Comment:
Shouldn't we be finding for the new tip in `otherTips` instead of picking the first entry?

Reply:
I suspect you're confusing this with the current _best_ chain tip, which is something different.

The tip being removed has already been found here (it was `entry.tip`).  The other tips, if any, are just other tips at the same height, so it doesn't really matter which one of them moves over to the static entry to take its place.

Reply:
For a concrete case, consider the following tips from the real main network:

```
$ dcrctl getchaintips | jq 'group_by(.height) | map(select(length==2)) | .[0]'
[
  {
    "height": 246490,
    "hash": "0000000000000001b2ed0133f165f3ce8300e01ac856c4bafef2d3085cc21dbb",
    "branchlen": 1,
    "status": "valid-headers"
  },
  {
    "height": 246490,
    "hash": "0000000000000001e9acb740a20c8fd4144af8e81dc8693d7acca8ab3c90eb35",
    "branchlen": 1,
    "status": "valid-headers"
  }
]
$ dcrctl getblockhash 246490
000000000000000220a2d5e28921b10c98fc659c6e3a39933f5a43a61b53c593
$ dcrctl getblockhash 246491
000000000000000073b5529e0e9c57dbc9df8a61de004620c4b10a7aced0dfbd
```

So, at some point, with this PR, the map of chaintips would have had an entry that looked like:

```Go
chainTips[246490] = chainTipEntry{
	tip: <node for 000000000000000220a2d5e28921b10c98fc659c6e3a39933f5a43a61b53c593>,
	otherTips: []*blockNode{
		<node for 0000000000000001b2ed0133f165f3ce8300e01ac856c4bafef2d3085cc21dbb>,
		<node for 0000000000000001e9acb740a20c8fd4144af8e81dc8693d7acca8ab3c90eb35>,
	},
}
```

Then, when the tip got extended ended up being `000000000000000220a2d5e28921b10c98fc659c6e3a39933f5a43a61b53c593`, it would go down this path, and modify the structure thusly:

```Go
chainTips[246490] = chainTipEntry{
	tip: <node for 0000000000000001b2ed0133f165f3ce8300e01ac856c4bafef2d3085cc21dbb>,
	otherTips: []*blockNode{
		<node for 0000000000000001e9acb740a20c8fd4144af8e81dc8693d7acca8ab3c90eb35>,
	},
}
chainTips[246491] = chainTipEntry{
	tip: <node for 000000000000000073b5529e0e9c57dbc9df8a61de004620c4b10a7aced0dfbd>,
	otherTips: nil,
}
```

Reply:
I see now, thanks for the detailed explanation.